interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,956 @@
+			continue
+		}
+		expired, err := matchingRule.Expired(currTime)
+		switch {
+		case err != nil:
+			// Issue with expiration, this should not occur
+			logger.Noticef("error while checking whether rule had expired: %v", err)
+			fallthrough
+		case expired:
+			continue
+		}
+		// Need to compare the path pattern variant, not the rule's path
+		// pattern, so that only variants which match are included,
+		// and the highest precedence variant can be computed.
+		matched, err := patterns.PathPatternMatches(variantStr, path)
+		if err != nil {
+			// Only possible error is ErrBadPattern, which should not occur
+			return false, err
+		}
+		if matched {
+			matchingVariants = append(matchingVariants, variantEntry.Variant)
+		}
+	}
+	if len(matchingVariants) == 0 {
+		return false, ErrNoMatchingRule
+	}
+	highestPrecedenceVariant, err := patterns.HighestPrecedencePattern(matchingVariants, path)
+	if err != nil {
+		return false, err
+	}
+	matchingEntry := variantMap[highestPrecedenceVariant.String()]
+	matchingID := matchingEntry.RuleID
+	matchingRule, err := rdb.ruleWithID(matchingID)
+	if err != nil {
+		// Database was left inconsistent, should not occur
+		return false, ErrRuleIDNotFound
+	}
+	return matchingRule.Outcome.AsBool()
+}
+
+// ruleWithIDForUser returns the rule with the given ID, if it exists, for the
+// given user. Otherwise, returns an error.
+//
+// The caller must ensure that the database lock is held.
+func (rdb *RuleDB) ruleWithIDForUser(user uint32, id prompting.IDType) (*Rule, error) {
+	rule, err := rdb.ruleWithID(id)
+	if err != nil {
+		return nil, err
+	}
+	if rule.User != user {
+		return nil, ErrUserNotAllowed
+	}
+	return rule, nil
+}
+
+// RuleWithID returns the rule with the given ID.
+// If the rule is not found, returns ErrRuleNotFound.
+// If the rule does not apply to the given user, returns ErrUserNotAllowed.
+func (rdb *RuleDB) RuleWithID(user uint32, id prompting.IDType) (*Rule, error) {
+	rdb.mutex.RLock()
+	defer rdb.mutex.RUnlock()
+	return rdb.ruleWithIDForUser(user, id)
+}
+
+// Creates a rule with the given information and adds it to the rule database.
+// If any of the given parameters are invalid, returns an error. Otherwise,
+// returns the newly-added rule, and saves the database to disk.
+func (rdb *RuleDB) AddRule(user uint32, snap string, iface string, constraints *prompting.Constraints, outcome prompting.OutcomeType, lifespan prompting.LifespanType, duration string) (*Rule, error) {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	newRule, err := rdb.populateNewRule(user, snap, iface, constraints, outcome, lifespan, duration)
+	if err != nil {
+		return nil, err
+	}
+	if needToSave, err, conflicts, conflictingPermission := rdb.addRuleToTree(newRule); err != nil {
+		if needToSave {
+			rdb.save() // discard any new error, which should not occur
+		}
+		return nil, fmt.Errorf("%w: conflicts: %+v, Permission: '%s'", err, conflicts, conflictingPermission)
```

Comment:
missed that earlier:
```suggestion
		return nil, fmt.Errorf("cannot add rule: %w: conflicts: %+v, permission: '%s'", err, conflicts, conflictingPermission)
```